server/testdriver/core/src/main/java/org/infinispan/server/test/core/rollingupgrade/RollingUpgradeHandler.java

```diff
@@ -0,0 +1,229 @@
+package org.infinispan.server.test.core.rollingupgrade;
+
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.server.Server;
+import org.infinispan.server.test.api.TestUser;
+import org.infinispan.server.test.core.ContainerInfinispanServerDriver;
+import org.infinispan.server.test.core.InfinispanServerTestConfiguration;
+import org.infinispan.server.test.core.ServerConfigBuilder;
+import org.infinispan.server.test.core.ServerRunMode;
+import org.infinispan.server.test.core.TestSystemPropertyNames;
+
+public class RollingUpgradeHandler {
+   private final RollingUpgradeConfiguration configuration;
+   private final Consumer<String> logConsumer;
+
+   private String toImageCreated;
+   private String fromImageCreated;
+
+   private ContainerInfinispanServerDriver fromDriver;
+   private ContainerInfinispanServerDriver toDriver;
+
+   private RemoteCacheManager remoteCacheManager;
+
+   private STATE currentState = STATE.NOT_STARTED;
+
+   public enum STATE {
+      /**
+       * The upgrade procedure has not been started yet
+       */
+      NOT_STARTED,
+      /**
+       * This state signals that all the old version nodes are running currently
+       */
+      OLD_RUNNING,
+      /**
+       * This state signals that one old node has been shut down. The remaining nodes may be a mixture of old and/or new
+       */
+      REMOVED_OLD,
+      /**
+       * This state signals that a new node was just added. The remaining nodes may be a mixture of old and/or new
+       */
+      ADDED_NEW,
+      /**
+       * This state signals that the upgrade completed and all old nodes have been removed and all new nodes are running
+       */
+      NEW_RUNNING,
+      /**
+       * The upgrade procedure encountered some type of error
+       */
+      ERROR
+   }
+
+   private RollingUpgradeHandler(RollingUpgradeConfiguration configuration) {
+      this.configuration = configuration;
+      this.logConsumer = configuration.logConsumer();
+   }
+
+   public String getToImageCreated() {
+      return toImageCreated;
+   }
+
+   public String getFromImageCreated() {
+      return fromImageCreated;
+   }
+
+   public ContainerInfinispanServerDriver getFromDriver() {
+      return fromDriver;
+   }
+
+   public ContainerInfinispanServerDriver getToDriver() {
+      return toDriver;
+   }
+
+   public RollingUpgradeConfiguration getConfiguration() {
+      return configuration;
+   }
+
+   public RemoteCacheManager getRemoteCacheManager() {
```

Comment:
Nit: Should we make it explicit the RCM connects to the "from" driver?

Reply:
There shouldn't be any difference technically as they are clustered together. So no need imo.